File: src/models/food-data.ts
// src/models/food-data.ts

/**
 * Nutritional Psychiatry Food Data Model
 * Core data model for the Nutritional Psychiatry Dataset
 */

// Base interfaces for reusable types
export interface MetadataObject {
  version: string;
  created: string;
  lastUpdated: string;
  sourceUrls: string[];
  sourceIds?: Record<string, string>;
  imageUrl?: string;
  tags: string[];
}

export interface DataQuality {
  completeness: number;
  overallConfidence: number;
  brainNutrientsSource?:
    | "usda_provided"
    | "openfoodfacts"
    | "literature_derived"
    | "ai_generated"
    | "expert_estimated";
  impactsSource?:
    | "direct_studies"
    | "literature_review"
    | "mechanism_inference"
    | "ai_generated";
  sourcePriority?: {
    standardNutrients?:
      | "usda"
      | "openfoodfacts"
      | "literature"
      | "ai_generated";
    brainNutrients?: "usda" | "openfoodfacts" | "literature" | "ai_generated";
    bioactiveCompounds?:
      | "usda"
      | "openfoodfacts"
      | "literature"
      | "ai_generated";
  };
}

export interface ServingInfo {
  servingSize: number;
  servingUnit: string;
  householdServing?: string;
}

export interface StandardNutrients {
  calories?: number;
  protein_g?: number;
  carbohydrates_g?: number;
  fat_g?: number;
  fiber_g?: number;
  sugars_g?: number;
  sugars_added_g?: number;
  calcium_mg?: number;
  iron_mg?: number;
  magnesium_mg?: number;
  phosphorus_mg?: number;
  potassium_mg?: number;
  sodium_mg?: number;
  zinc_mg?: number;
  copper_mg?: number;
  manganese_mg?: number;
  selenium_mcg?: number;
  vitamin_c_mg?: number;
  vitamin_a_iu?: number;
}

export interface Omega3 {
  total_g?: number;
  epa_mg?: number;
  dha_mg?: number;
  ala_mg?: number;
  confidence?: number;
}

export interface BrainNutrients {
  tryptophan_mg?: number;
  tyrosine_mg?: number;
  vitamin_b6_mg?: number;
  folate_mcg?: number;
  vitamin_b12_mcg?: number;
  vitamin_d_mcg?: number;
  magnesium_mg?: number;
  zinc_mg?: number;
  iron_mg?: number;
  selenium_mcg?: number;
  choline_mg?: number;
  omega3?: Omega3;
}

export interface BioactiveCompounds {
  polyphenols_mg?: number;
  flavonoids_mg?: number;
  anthocyanins_mg?: number;
  carotenoids_mg?: number;
  probiotics_cfu?: number;
  prebiotic_fiber_g?: number;
}

export interface ResearchSupport {
  citation: string;
  doi?: string;
  url?: string;
  study_type?: string;
  year?: number;
}

export interface MentalHealthImpact {
  impact_type:
    | "mood_elevation"
    | "mood_depression"
    | "anxiety_reduction"
    | "anxiety_increase"
    | "cognitive_enhancement"
    | "cognitive_decline"
    | "energy_increase"
    | "energy_decrease"
    | "stress_reduction"
    | "sleep_improvement"
    | "gut_health_improvement";
  direction: "positive" | "negative" | "neutral" | "mixed";
  mechanism: string;
  strength: number;
  confidence: number;
  time_to_effect?: "acute" | "cumulative" | "both_acute_and_cumulative";
  research_context?: string;
  research_support?: ResearchSupport[];
  notes?: string;
}

export interface NutrientInteraction {
  interaction_id: string;
  nutrients_involved: string[];
  interaction_type:
    | "synergistic"
    | "antagonistic"
    | "required_cofactor"
    | "protective"
    | "inhibitory";
  pathway?: string;
  mechanism: string;
  mental_health_relevance?: string;
  confidence: number;
  research_support?: ResearchSupport[];
  foods_demonstrating?: string[];
}

export interface CircadianFactor {
  factor: string;
  effects: string[];
  relevant_to: string[];
  confidence: number;
  citations?: string[];
}

export interface CircadianEffects {
  description?: string;
  factors: CircadianFactor[];
}

export interface FoodCombination {
  combination: string;
  effects: string[];
  relevant_to: string[];
  confidence: number;
}

export interface PreparationEffect {
  method: string;
  effects: string[];
  relevant_to: string[];
  confidence: number;
}

export interface ContextualFactors {
  circadian_effects?: CircadianEffects;
  food_combinations?: FoodCombination[];
  preparation_effects?: PreparationEffect[];
}

export interface PopulationVariation {
  population: string;
  description: string;
  variations: {
    nutrient: string;
    effect: string;
    mechanism?: string;
    impact_modifier?: number;
    recommendations?: string[];
    confidence: number;
    citations?: string[];
  }[];
}

export interface DietaryPattern {
  pattern_name:
    | "mediterranean"
    | "western"
    | "dash"
    | "mind"
    | "vegetarian"
    | "vegan"
    | "ketogenic"
    | "low_fodmap";
  pattern_contribution:
    | "key_component"
    | "supportive"
    | "occasional"
    | "limited"
    | "avoided";
  mental_health_relevance?: string;
}

export interface InflammatoryIndex {
  value: number;
  confidence: number;
  calculation_method:
    | "dietary_inflammatory_index"
    | "empirical_dietary_index"
    | "expert_estimate";
  citations: string[];
}

export interface NeuralTarget {
  pathway: string;
  effect: "upregulation" | "downregulation" | "modulation" | "protection";
  confidence: number;
  mechanisms?: string[];
  mental_health_relevance?: string;
}

/**
 * Main FoodData interface representing the complete schema
 */
export interface FoodData {
  food_id: string;
  name: string;
  description?: string;
  category: string;
  serving_info?: ServingInfo;
  standard_nutrients: StandardNutrients;
  brain_nutrients?: BrainNutrients;
  bioactive_compounds?: BioactiveCompounds;
  mental_health_impacts?: MentalHealthImpact[];
  nutrient_interactions?: NutrientInteraction[];
  contextual_factors?: ContextualFactors;
  population_variations?: PopulationVariation[];
  dietary_patterns?: DietaryPattern[];
  inflammatory_index?: InflammatoryIndex;
  neural_targets?: NeuralTarget[];
  data_quality: DataQuality;
  metadata: MetadataObject;
}

/**
 * Type guards for runtime type checking
 */
export function isFoodData(obj: any): obj is FoodData {
  return (
    obj &&
    typeof obj.food_id === "string" &&
    typeof obj.name === "string" &&
    typeof obj.category === "string" &&
    typeof obj.standard_nutrients === "object"
  );
}

File: src/models/food-data.py
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass, field, asdict
from datetime import datetime

@dataclass
class ServingInfo:
    serving_size: float = 100.0
    serving_unit: str = "g"
    household_serving: Optional[str] = None

@dataclass
class StandardNutrients:
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    carbohydrates_g: Optional[float] = None
    fat_g: Optional[float] = None
    fiber_g: Optional[float] = None
    sugars_g: Optional[float] = None
    sugars_added_g: Optional[float] = None
    calcium_mg: Optional[float] = None
    iron_mg: Optional[float] = None
    magnesium_mg: Optional[float] = None
    phosphorus_mg: Optional[float] = None
    potassium_mg: Optional[float] = None
    sodium_mg: Optional[float] = None
    zinc_mg: Optional[float] = None
    copper_mg: Optional[float] = None
    manganese_mg: Optional[float] = None
    selenium_mcg: Optional[float] = None
    vitamin_c_mg: Optional[float] = None
    vitamin_a_iu: Optional[float] = None

@dataclass
class Omega3:
    total_g: Optional[float] = None
    epa_mg: Optional[float] = None
    dha_mg: Optional[float] = None
    ala_mg: Optional[float] = None
    confidence: Optional[float] = None

@dataclass
class BrainNutrients:
    tryptophan_mg: Optional[float] = None
    tyrosine_mg: Optional[float] = None
    vitamin_b6_mg: Optional[float] = None
    folate_mcg: Optional[float] = None
    vitamin_b12_mcg: Optional[float] = None
    vitamin_d_mcg: Optional[float] = None
    magnesium_mg: Optional[float] = None
    zinc_mg: Optional[float] = None
    iron_mg: Optional[float] = None
    selenium_mcg: Optional[float] = None
    choline_mg: Optional[float] = None
    omega3: Optional[Omega3] = None

@dataclass
class BioactiveCompounds:
    polyphenols_mg: Optional[float] = None
    flavonoids_mg: Optional[float] = None
    anthocyanins_mg: Optional[float] = None
    carotenoids_mg: Optional[float] = None
    probiotics_cfu: Optional[float] = None
    prebiotic_fiber_g: Optional[float] = None

@dataclass
class ResearchSupport:
    citation: str
    doi: Optional[str] = None
    url: Optional[str] = None
    study_type: Optional[str] = None
    year: Optional[int] = None

@dataclass
class MentalHealthImpact:
    impact_type: str  # One of mood_elevation, mood_depression, etc.
    direction: str  # positive, negative, neutral, mixed
    mechanism: str
    strength: float
    confidence: float
    time_to_effect: Optional[str] = None  # acute, cumulative, both_acute_and_cumulative
    research_context: Optional[str] = None
    research_support: List[ResearchSupport] = field(default_factory=list)
    notes: Optional[str] = None

@dataclass
class DataQuality:
    completeness: float
    overall_confidence: float
    brain_nutrients_source: Optional[str] = None
    impacts_source: Optional[str] = None
    source_priority: Dict[str, str] = field(default_factory=dict)

@dataclass
class Metadata:
    version: str
    created: str
    last_updated: str
    source_urls: List[str] = field(default_factory=list)
    source_ids: Dict[str, str] = field(default_factory=dict)
    image_url: Optional[str] = None
    tags: List[str] = field(default_factory=list)

@dataclass
class InflammatoryIndex:
    value: float
    confidence: float
    calculation_method: str
    citations: List[str] = field(default_factory=list)

@dataclass
class FoodData:
    food_id: str
    name: str
    category: str
    standard_nutrients: StandardNutrients
    data_quality: DataQuality
    metadata: Metadata
    description: Optional[str] = None
    serving_info: Optional[ServingInfo] = None
    brain_nutrients: Optional[BrainNutrients] = None
    bioactive_compounds: Optional[BioactiveCompounds] = None
    mental_health_impacts: List[MentalHealthImpact] = field(default_factory=list)
    inflammatory_index: Optional[InflammatoryIndex] = None
    
    def to_dict(self) -> Dict:
        """Convert dataclass to dictionary."""
        return asdict(self)
    
    @classmethod
    def from_dict(cls, data: Dict) -> 'FoodData':
        """Create FoodData instance from dictionary."""
        # Handle nested dataclasses
        standard_nutrients = StandardNutrients(**data.get('standard_nutrients', {}))
        
        data_quality = DataQuality(**data.get('data_quality', {
            'completeness': 0.0,
            'overall_confidence': 5.0
        }))
        
        metadata = Metadata(**data.get('metadata', {
            'version': '0.1.0',
            'created': datetime.now().isoformat(),
            'last_updated': datetime.now().isoformat(),
        }))
        
        serving_info = None
        if 'serving_info' in data:
            serving_info = ServingInfo(**data['serving_info'])
        
        brain_nutrients = None
        if 'brain_nutrients' in data:
            brain_data = data['brain_nutrients'].copy()
            if 'omega3' in brain_data:
                brain_data['omega3'] = Omega3(**brain_data['omega3'])
            brain_nutrients = BrainNutrients(**brain_data)
        
        bioactive_compounds = None
        if 'bioactive_compounds' in data:
            bioactive_compounds = BioactiveCompounds(**data['bioactive_compounds'])
        
        mental_health_impacts = []
        for impact in data.get('mental_health_impacts', []):
            research_support = []
            for support in impact.get('research_support', []):
                research_support.append(ResearchSupport(**support))
            
            impact_copy = impact.copy()
            impact_copy['research_support'] = research_support
            mental_health_impacts.append(MentalHealthImpact(**impact_copy))
        
        inflammatory_index = None
        if 'inflammatory_index' in data:
            inflammatory_index = InflammatoryIndex(**data['inflammatory_index'])
        
        return cls(
            food_id=data.get('food_id', ''),
            name=data.get('name', ''),
            category=data.get('category', ''),
            description=data.get('description'),
            standard_nutrients=standard_nutrients,
            serving_info=serving_info,
            brain_nutrients=brain_nutrients,
            bioactive_compounds=bioactive_compounds,
            mental_health_impacts=mental_health_impacts,
            inflammatory_index=inflammatory_index,
            data_quality=data_quality,
            metadata=metadata
        )